homeassistant/components/cloudflare/__init__.py

```diff
@@ -97,22 +93,52 @@ async def _async_update_cloudflare(
 ) -> None:
     _LOGGER.debug("Starting update for zone %s", dns_zone["name"])
 
-    records = await client.list_dns_records(zone_id=dns_zone["id"], type="A")
-    _LOGGER.debug("Records: %s", records)
+    old_records = await client.list_dns_records(zone_id=dns_zone["id"])
+    _LOGGER.debug("Records: %s", old_records)
 
     session = async_get_clientsession(hass, family=socket.AF_INET)
-    location_info = await async_detect_location_info(session)
-
-    if not location_info or not is_ipv4_address(location_info.ip):
-        raise HomeAssistantError("Could not get external IPv4 address")
+    session_ipv6 = async_get_clientsession(hass, family=socket.AF_INET6)
 
-    filtered_records = [
-        record
-        for record in records
-        if record["name"] in target_records and record["content"] != location_info.ip
-    ]
-
-    if len(filtered_records) == 0:
```

Comment:
Keep this filtering, but move it before the location info calls.

Reply:
That way you can call the correct handler based on the current type.

Reply:
The logic hasn't been removed. Based on suggestions from other reviewers, the variable name was changed, which I now understand. 😊

Reply:
Right, now you are looking over it potentially 2 times, where only 1 is needed.
You are also calling the locations service 2 times on each invocation, where only one of these are needed (depending on the record type).

In short:
- Only use v4 location service if the record type is A
- Only use v6 location service if the record type is AAA

Reply:
Hi @ludeeus,

Thank you for your feedback! Based on your suggestion, I’ve optimized the logic to ensure async_detect_location_info is called only once per record type (IPv4/IPv6), and the filtering is done in one pass. This avoids redundant service calls and improves efficiency. Does this look better to you?

Updated code:

```
    records_to_be_updated = []

    for record in old_records:
        if record["name"] not in target_records:
            continue
        if record["type"] == "A":
            _LOGGER.debug("Found target record: %s", record["name"])
            session = async_get_clientsession(hass, family=socket.AF_INET)
            location_info = await async_detect_location_info(session)
            if location_info and is_ipv4_address(location_info.ip):
                _LOGGER.debug("IPv4 address detected: %s", location_info.ip)
                if record["content"] != location_info.ip:
                    _LOGGER.info(
                        "IPv4 address change detected for record: %s,will DNS entry from %s to %s",
                        record["name"],
                        record["content"],
                        location_info.ip,
                    )
                    record["content"] = location_info.ip
                    records_to_be_updated.append(record)
        if record["type"] == "AAAA":
            _LOGGER.debug("Found target record: %s", record["name"])
            session_ipv6 = async_get_clientsession(hass, family=socket.AF_INET6)
            location_info_v6 = await async_detect_location_info(session_ipv6)
            if location_info_v6 and is_ipv6_address(location_info_v6.ip):
                _LOGGER.debug("IPv6 address detected: %s", location_info_v6.ip)
                if record["content"] != location_info_v6.ip:
                    _LOGGER.info(
                        "IPv6 address change detected for record: %s,will DNS entry from %s to %s",
                        record["name"],
                        record["content"],
                        location_info_v6.ip,
                    )
                    record["content"] = location_info_v6.ip
                    records_to_be_updated.append(record)
```

Would love to hear your thoughts on this approach. 😊

Reply:
Looks better 👍 
Just one note; Instead of looping like this, extract the correct record form the list, and act on that object (if found).